src/adtConstructors.ml

```diff
@@ -0,0 +1,303 @@
+        indent (separate (!^ ";" ^^ newline) (fields |> List.map (fun (x, typ) ->
+            nest (Name.to_coq x ^^ !^ ":" ^^ Type.to_coq None None typ)))) ^^
+        !^ "}." ^^
+        begin match typ_args with
+          | [] -> empty
+          | _ :: _ ->
+            newline ^^ !^ "Arguments" ^^ !^ "record" ^^ !^ ":" ^^
+            !^ "clear" ^^ !^ "implicits" ^-^ !^ "."
+        end ^^
+        newline ^^
+        begin if with_with then
+            separate newline (fields |> List.map (fun (name, _) ->
+                let prefixed_typ_args =
+                  typ_args |> List.map (fun typ_arg ->
+                      Name.to_coq (Name.prefix_by_t typ_arg)
+                    ) in
+                let record_typ =
+                  nest (separate space (!^ "record" :: prefixed_typ_args)) in
+                nest (
+                  !^ "Definition" ^^ Name.to_coq (Name.prefix_by_with name) ^^
+                  begin match typ_args with
+                    | [] -> empty
+                    | _ :: _ -> braces (nest (separate space prefixed_typ_args))
+                  end ^^
+                  Name.to_coq name ^^
+                  nest (parens (!^ "r" ^^ !^ ":" ^^ record_typ)) ^-^
+                  !^ " :=" ^^ newline ^^
+                  indent @@ nest (
+                    !^ "Build" ^^
+                    separate space prefixed_typ_args ^^
+                    separate space (fields |> List.map (fun (name', _) ->
+                        nest (
+                          if Name.equal name name' then
+                            Name.to_coq name
+                          else
+                            !^ "r" ^-^ !^ ".(" ^-^ Name.to_coq name' ^-^ !^ ")"
+                        )
+                      )) ^-^ !^ "."
+                  )
+                )
+              )) ^^ newline
+          else
+            empty
+        end
+      ) ^^
+      !^ "End" ^^ Name.to_coq module_name ^-^ !^ "." ^^ newline ^^
+      nest (
+        !^ "Definition" ^^ Name.to_coq typ_name ^^ !^ ":=" ^^
+        Name.to_coq module_name ^-^ !^ ".record" ^-^ !^ "."
+      )
+    )
+
+
+  let to_coq (record_skeleton : t) : SmartPrint.t =
+    let { fields; module_name; typ_name } = record_skeleton in
+    to_coq_record module_name typ_name fields (fields |>
+                                               List.map (fun field -> (field, Type.Variable field))
+                                              ) true
+end
+
+type ret_typ =
+  | Variant of AdtParameters.t
+  | Tagged of Type.t list
+
+let ret_is_tagged : ret_typ -> bool = function
+  | Variant _ -> false
+  | Tagged _ -> true
+
+(** The constructors of an inductive type, either in a GADT or non-GADT form. *)
+(** [constructor_name]: forall [typ_vars], [param_typs] -> t [res_typ_params] *)
+type item = {
+  constructor_name : Name.t;
+  param_typs : Type.t list; (** The parameters of the constructor. *)
+  res_typ_params : ret_typ;
+  res_typ_length : int ;
+  is_tagged : bool;
```

Comment:
`res_typ_length` definitely can. `is_tagged` cannot because it is computed strictly from the attributes, so in one way or another you have to have a way to retrieve the information from attributes.

My thought process to add the length as well is because I thought it would be preferable to use a little more memory than having to recompute this same information many times.

Reply:
Ah, OK for `is_tagged`. I think for `coq-of-ocaml` efficiency is not an issue (at least for now and for most of the code). By invariant I mean that there is an invariant relating `res_typ_params` and `res_typ_params`, namely `List.length res_typ_params = res_typ_length`, which could be broken by latter refactorings.